### PR TITLE
Fix memory allocation problems in e_loader_attic.c

### LINKED ISSUE #15115: [BUG] A double free bug in engines/e_loader_attic.c
**File**: engines/e_loader_attic.c
**Bug Function:** try_decode_X509Certificate
Version: Git-master 

**Description:**
My analyzer reported a double free bug.

1. In function `try_decode_X509Certificate`,  at line 785, `d2i_X509_AUX(&cert, &blob, len)`  is called.
2. When error happens in `d2i_X509_AUX()`,  the callee will return `NULL` and the `cert` is set to `NULL`. 
Then `d2i_X509(&cert, &blob, len)` at line 786 will be called.
3. Inside `d2i_X509()`, if the condition `!ossl_x509v3_cache_extensions(cert)` at line 133 is true, the `cert` will be **freed on the first time.** 
4. Back to `try_decode_X509Certificate`, the `cert` will be **freed again** at line 792.

Note that, if the first error caused by limited available memory, the after errors can also be easy triggerd.

**Code Description:**
```
File: engines/e_loader_attic.c
static OSSL_STORE_INFO *try_decode_X509Certificate(const char *pem_name,)
{
    store_info = NULL;
    ...
               //in err, d2i_X509_AUX()  set cert = NULL, return NULL
785:    if ((d2i_X509_AUX(&cert, &blob, len)) != NULL
786:        || (ignore_trusted && (d2i_X509(&cert, &blob, len)) != NULL)) {
                            //Then d2i_X509() is called, and free cert in error.
        *matchcount = 1;
        store_info = OSSL_STORE_INFO_new_CERT(cert);
    }

    if (store_info == NULL)
792:    X509_free(cert);  // cert is freed again !

    return store_info;
}

File: crypto/x509/x_x509.c
X509 *d2i_X509(X509 **a, const unsigned char **in, long len)
{
    X509 *cert = NULL;
    int free_on_error = a != NULL && *a == NULL; // free_on_error is true

    cert = (X509 *)ASN1_item_d2i((ASN1_VALUE **)a, in, len, (X509_it()));
    // set cert = *a

    if (cert != NULL && a != NULL) { // This condition is true if the above callee run successfully.
133:    if (!x509v3_cache_extensions(cert)) {  // if failed.
            if (free_on_error)
                X509_free(cert);  // *a is freed.
            cert = NULL;
        }
    }
    return cert;
}

```

### PR DESCRIPTION
**Pull request by paulidale**
There are a double free and a use after free problem.
This fixes both.
Refer to the PRs for a description of what the problem is.

Fixes #15115 and #15116

### CODE REVIEW

File: engines/e_loader_attic.c [hunk 1]
```diff
@@ -783,8 +784,12 @@ static OSSL_STORE_INFO *try_decode_X509Certificate(const char *pem_name,
     if (cert == NULL)
         return NULL;
 
-    if ((d2i_X509_AUX(&cert, &blob, len)) != NULL
-        || (ignore_trusted && (d2i_X509(&cert, &blob, len)) != NULL)) {
+    create_cert = d2i_X509_AUX(&cert, &blob, len) != NULL;
+    if (!create_cert && ignore_trusted)
+        /* d2i_X509 will set cert but free the memory on error */
```
Comment (levitte) [inline engines/e_loader_attic.c:789]: Argh!  That's a bug that should be fixed *there*

Comment (paulidale) [reply engines/e_loader_attic.c:789]: Quite a bit of code relies on the current behaviour.  Tests go crazy.

Comment (levitte) [reply engines/e_loader_attic.c:789]:
``` diff
diff --git a/crypto/x509/x_x509.c b/crypto/x509/x_x509.c
index 529d701bbb..ea664f8214 100644
--- a/crypto/x509/x_x509.c
+++ b/crypto/x509/x_x509.c
@@ -134,6 +134,9 @@ X509 *d2i_X509(X509 **a, const unsigned char **in, long len)
             if (free_on_error)
                 X509_free(cert);
             cert = NULL;
+            /* Prevent *a from pointing at freed memory */
+            if (a != NULL)
+                *a = NULL;
         }
     }
     return cert;
```

Comment (paulidale) [reply engines/e_loader_attic.c:789]: That's exactly the change I tried.  It broke things.

Comment (levitte) [reply engines/e_loader_attic.c:789]: > Quite a bit of code relies on the current behaviour. Tests go crazy.

Really?  I gotta check that out...

Comment (levitte) [reply engines/e_loader_attic.c:789]: (it could be that there are faults in the tests)

Comment (paulidale) [reply engines/e_loader_attic.c:789]: A neater change is just: `*a = cert = NULL;`  `a` cannot be NULL at this point.

Comment (paulidale) [reply engines/e_loader_attic.c:789]: Figure it out I think.  Fix coming soon.

Comment (paulidale) [reply engines/e_loader_attic.c]: Pushed.  The fix turned out nice and simple.

**Comment by paulidale** [comment]
I think the sanitizer failure is a different problem.

**Review by levitte — APPROVED** [review]

**Comment by paulidale** [comment]
Commit message updated to match PR.

**Review by levitte — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master, thanks for the comments.

**Pull request closed without merging**